File: core/src/main/java/meshu/core/l3/Ops.java
package meshu.core.l3;

import java.util.ArrayList;
import java.util.List;

/**
 * Operation wire formats (PROTOCOL.md §8) — builders and parsers for the PoC
 * subset (POC-WALLET.md §3): HELLO, KEYSETS, KEYS, CHECKSTATE, SWAP, RESTORE,
 * MINT_INFO, ERROR.
 *
 * <p>Each op has a {@code request(...)} producing an {@link Envelope} and a
 * {@code parse(Response)} decoding the response {@link Envelope}. Decoders
 * ignore unknown trailing elements (§5.2) for forward compatibility.
 */
public final class Ops {

    private Ops() {
    }

    private static Cbor.Value.Uint u(Cbor.Value v, String what) {
        if (!(v instanceof Cbor.Value.Uint x)) {
            throw new IllegalArgumentException("expected uint for " + what);
        }
        return x;
    }

    private static byte[] b(Cbor.Value v, String what) {
        if (!(v instanceof Cbor.Value.Bytes x)) {
            throw new IllegalArgumentException("expected bytes for " + what);
        }
        return x.value();
    }

    private static List<Cbor.Value> arr(Cbor.Value v, String what) {
        if (!(v instanceof Cbor.Value.Array x)) {
            throw new IllegalArgumentException("expected array for " + what);
        }
        return x.items();
    }

    // ============================================================ HELLO (§8.1)

    /**
     * req: {@code [ 0x0D, l1_version, wallet_pubkey, [supported_ops...] ]}
     */
    public static Envelope helloRequest(int l1Version, byte[] walletPubkey32, int[] supportedOps) {
        if (walletPubkey32.length != 32) {
            throw new IllegalArgumentException("wallet_pubkey must be 32 bytes");
        }
        List<Cbor.Value> ops = new ArrayList<>();
        for (int op : supportedOps) {
            ops.add(Cbor.uint(op));
        }
        return Envelope.of(Op.HELLO,
                Cbor.uint(l1Version),
                Cbor.bytes(walletPubkey32),
                Cbor.array(ops));
    }

    /**
     * resp: {@code [ 0x8D, l1_version, gateway_pubkey, [mint_urls...],
     * [nut19_cached_paths...], max_msg, server_time ]}
     */
    public record HelloResponse(
            int l1Version,
            byte[] gatewayPubkey,
            List<String> mintUrls,
            List<String> nut19CachedPaths,
            long maxMsg,
            long serverTime) {
    }

    public static HelloResponse parseHelloResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.HELLO));
        List<Cbor.Value> f = env.fields();
        return new HelloResponse(
                u(f.get(0), "l1_version").intValue(),
                b(f.get(1), "gateway_pubkey"),
                textList(f.get(2), "mint_urls"),
                textList(f.get(3), "nut19_cached_paths"),
                u(f.get(4), "max_msg").longValue(),
                u(f.get(5), "server_time").longValue());
    }

    public static Envelope helloResponse(int l1Version, byte[] gatewayPubkey32,
                                         List<String> mintUrls, List<String> nut19CachedPaths,
                                         long maxMsg, long serverTime) {
        return Envelope.of(Op.responseOf(Op.HELLO),
                Cbor.uint(l1Version),
                Cbor.bytes(gatewayPubkey32),
                Cbor.array(mintUrls.stream().map(Cbor::text).map(v -> (Cbor.Value) v).toList()),
                Cbor.array(nut19CachedPaths.stream().map(Cbor::text).map(v -> (Cbor.Value) v).toList()),
                Cbor.uint(maxMsg),
                Cbor.uint(serverTime));
    }

    // ============================================================ KEYSETS (§8.10)

    /** req: {@code [ 0x0A, mint_handle ]} */
    public static Envelope keysetsRequest(int mintHandle) {
        return Envelope.of(Op.KEYSETS, Cbor.uint(mintHandle));
    }

    /**
     * resp: {@code [ 0x8A, [ [ s_id, unit, active, input_fee_ppk, final_expiry ], ... ] ]}
     * Response order defines the keyset handles: handle i is entry i.
     */
    public record KeysetEntry(
            byte[] shortId,
            String unit,
            boolean active,
            long inputFeePpk,
            long finalExpiry) {
    }

    public static List<KeysetEntry> parseKeysetsResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.KEYSETS));
        List<Cbor.Value> rows = arr(env.fields().get(0), "keysets");
        List<KeysetEntry> out = new ArrayList<>();
        for (Cbor.Value row : rows) {
            List<Cbor.Value> e = arr(row, "keyset entry");
            out.add(new KeysetEntry(
                    b(e.get(0), "s_id"),
                    Unit.fromCbor(e.get(1)),
                    u(e.get(2), "active").longValue() != 0,
                    u(e.get(3), "input_fee_ppk").longValue(),
                    u(e.get(4), "final_expiry").longValue()));
        }
        return out;
    }

    public static Envelope keysetsResponse(List<KeysetEntry> entries) {
        List<Cbor.Value> rows = new ArrayList<>();
        for (KeysetEntry e : entries) {
            rows.add(Cbor.array(
                    Cbor.bytes(e.shortId()),
                    Unit.toCbor(e.unit()),
                    Cbor.uint(e.active() ? 1 : 0),
                    Cbor.uint(e.inputFeePpk()),
                    Cbor.uint(e.finalExpiry())));
        }
        return Envelope.of(Op.responseOf(Op.KEYSETS), Cbor.array(rows));
    }

    // ============================================================ KEYS (§8.11)

    /** req: {@code [ 0x0B, keyset_handle, from_exp?, to_exp? ]} */
    public static Envelope keysRequest(int keysetHandle, Integer fromExp, Integer toExp) {
        List<Cbor.Value> fields = new ArrayList<>();
        fields.add(Cbor.uint(keysetHandle));
        fields.add(fromExp == null ? Cbor.Value.Null.INSTANCE : Cbor.uint(fromExp));
        fields.add(toExp == null ? Cbor.Value.Null.INSTANCE : Cbor.uint(toExp));
        return new Envelope(Op.KEYS, fields);
    }

    /** resp: {@code [ 0x8B, s_id, keys_blob ]} */
    public record KeysResponse(byte[] shortId, List<byte[]> keys) {
    }

    public static KeysResponse parseKeysResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.KEYS));
        List<Cbor.Value> f = env.fields();
        return new KeysResponse(
                b(f.get(0), "s_id"),
                PackedBlobs.unpackKeys(b(f.get(1), "keys_blob")));
    }

    public static Envelope keysResponse(byte[] shortId, List<byte[]> keys) {
        return Envelope.of(Op.responseOf(Op.KEYS),
                Cbor.bytes(shortId),
                Cbor.bytes(PackedBlobs.packKeys(keys)));
    }

    // ============================================================ CHECKSTATE (§8.9)

    /** req: {@code [ 0x09, y_blob ]} */
    public static Envelope checkstateRequest(List<byte[]> ys) {
        return Envelope.of(Op.CHECKSTATE, Cbor.bytes(PackedBlobs.packYs(ys)));
    }

    /** Proof states (§8.9): 2 bits per proof, LSB-first. */
    public static final int STATE_UNSPENT = 0;
    public static final int STATE_PENDING = 1;
    public static final int STATE_SPENT = 2;

    /** resp: {@code [ 0x89, packed_states ]} */
    public static int[] parseCheckstateResponse(Envelope env, int proofCount) {
        requireOp(env, Op.responseOf(Op.CHECKSTATE));
        byte[] packed = b(env.fields().get(0), "packed_states");
        return unpackStates(packed, proofCount);
    }

    public static Envelope checkstateResponse(int[] states) {
        return Envelope.of(Op.responseOf(Op.CHECKSTATE), Cbor.bytes(packStates(states)));
    }

    /** Pack 2-bit states, LSB-first (§8.9). */
    public static byte[] packStates(int[] states) {
        byte[] out = new byte[(states.length * 2 + 7) / 8];
        for (int i = 0; i < states.length; i++) {
            int s = states[i];
            if (s < 0 || s > 2) {
                throw new IllegalArgumentException("state out of range: " + s);
            }
            out[(i * 2) >> 3] |= (byte) (s << ((i * 2) & 7));
        }
        return out;
    }

    public static int[] unpackStates(byte[] packed, int count) {
        int[] out = new int[count];
        for (int i = 0; i < count; i++) {
            out[i] = (packed[(i * 2) >> 3] >> ((i * 2) & 7)) & 0x3;
        }
        return out;
    }

    // ============================================================ SWAP (§8.13)

    /**
     * req: {@code [ 0x08, keyset_handle, inputs_blob, outputs_blob ]}
     */
    public static Envelope swapRequest(int keysetHandle,
                                       List<PackedBlobs.Proof> inputs,
                                       List<PackedBlobs.Output> outputs) {
        return Envelope.of(Op.SWAP,
                Cbor.uint(keysetHandle),
                Cbor.bytes(PackedBlobs.packProofs(inputs)),
                Cbor.bytes(PackedBlobs.packOutputs(outputs)));
    }

    /** resp: {@code [ 0x88, signatures_blob ]} */
    public static List<byte[]> parseSwapResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.SWAP));
        return PackedBlobs.unpackSignatures(b(env.fields().get(0), "signatures_blob"));
    }

    public static Envelope swapResponse(List<byte[]> signatures) {
        return Envelope.of(Op.responseOf(Op.SWAP), Cbor.bytes(PackedBlobs.packSignatures(signatures)));
    }

    // ============================================================ RESTORE (§8.8)

    /**
     * req: {@code [ 0x07, keyset_handle, counter_start, count, outputs_blob ]}
     */
    public static Envelope restoreRequest(int keysetHandle, long counterStart, int count,
                                          List<PackedBlobs.Output> outputs) {
        return Envelope.of(Op.RESTORE,
                Cbor.uint(keysetHandle),
                Cbor.uint(counterStart),
                Cbor.uint(count),
                Cbor.bytes(PackedBlobs.packOutputs(outputs)));
    }

    /**
     * resp: {@code [ 0x87, hit_bitmap, signatures_blob ]}
     * hit_bitmap bit i set → counter (counter_start + i) was signed; signatures
     * for set bits only, ascending.
     */
    public record RestoreResponse(byte[] hitBitmap, List<byte[]> signatures) {
    }

    public static RestoreResponse parseRestoreResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.RESTORE));
        List<Cbor.Value> f = env.fields();
        return new RestoreResponse(
                b(f.get(0), "hit_bitmap"),
                PackedBlobs.unpackSignatures(b(f.get(1), "signatures_blob")));
    }

    public static Envelope restoreResponse(byte[] hitBitmap, List<byte[]> signatures) {
        return Envelope.of(Op.responseOf(Op.RESTORE),
                Cbor.bytes(hitBitmap),
                Cbor.bytes(PackedBlobs.packSignatures(signatures)));
    }

    // ============================================================ MINT_INFO (§8.12)

    /** req: {@code [ 0x0C, mint_handle ]} */
    public static Envelope mintInfoRequest(int mintHandle) {
        return Envelope.of(Op.MINT_INFO, Cbor.uint(mintHandle));
    }

    /**
     * resp: {@code [ 0x8C, name?, [ nut_numbers... ], min_amount, max_amount,
     * [ method_unit_pairs... ], nut19_ttl? ]}
     */
    public record MintInfoResponse(
            String name,
            List<Integer> nutNumbers,
            long minAmount,
            long maxAmount,
            List<String> methodUnitPairs,
            Long nut19Ttl) {
    }

    public static MintInfoResponse parseMintInfoResponse(Envelope env) {
        requireOp(env, Op.responseOf(Op.MINT_INFO));
        List<Cbor.Value> f = env.fields();
        String name = f.get(0) instanceof Cbor.Value.Text t ? t.value() : null;
        List<Integer> nuts = new ArrayList<>();
        for (Cbor.Value v : arr(f.get(1), "nut_numbers")) {
            nuts.add(u(v, "nut").intValue());
        }
        List<String> pairs = textList(f.get(4), "method_unit_pairs");
        Long ttl = f.size() > 5 && f.get(5) instanceof Cbor.Value.Uint t ? t.longValue() : null;
        return new MintInfoResponse(name, nuts,
                u(f.get(2), "min_amount").longValue(),
                u(f.get(3), "max_amount").longValue(),
                pairs, ttl);
    }

    // ============================================================ ERROR (§8.14, §10)

    /**
     * resp: {@code [ 0xFE, code, detail_kind, detail? ]}
     * detail_kind: 0 = none, 1 = text, 2 = bytes.
     */
    public record ErrorResponse(long code, int detailKind, byte[] detailBytes, String detailText) {
    }

    public static Envelope error(long code) {
        return Envelope.of(Op.ERROR, Cbor.uint(code), Cbor.uint(0));
    }

    public static Envelope errorText(long code, String detail) {
        return Envelope.of(Op.ERROR, Cbor.uint(code), Cbor.uint(1), Cbor.text(detail));
    }

    public static ErrorResponse parseError(Envelope env) {
        requireOp(env, Op.ERROR);
        List<Cbor.Value> f = env.fields();
        long code = u(f.get(0), "code").longValue();
        int kind = u(f.get(1), "detail_kind").intValue();
        byte[] bytes = null;
        String text = null;
        if (kind == 1 && f.size() > 2) {
            text = ((Cbor.Value.Text) f.get(2)).value();
        } else if (kind == 2 && f.size() > 2) {
            bytes = b(f.get(2), "detail");
        }
        return new ErrorResponse(code, kind, bytes, text);
    }

    // ============================================================ helpers

    private static void requireOp(Envelope env, int expected) {
        if (env.opcode() != expected) {
            throw new IllegalArgumentException(
                    "expected opcode 0x%02x, got 0x%02x".formatted(expected, env.opcode()));
        }
    }

    private static List<String> textList(Cbor.Value v, String what) {
        List<String> out = new ArrayList<>();
        for (Cbor.Value item : arr(v, what)) {
            if (!(item instanceof Cbor.Value.Text t)) {
                throw new IllegalArgumentException("expected text in " + what);
            }
            out.add(t.value());
        }
        return out;
    }
}

File: wallet/src/l3/ops.ts
/**
 * §8 op wire-format builders/parsers for the PoC subset — TS port of
 * core/Ops.java. Decoders ignore unknown trailing elements (§5.2).
 */
import {
  array,
  asArray,
  asBytes,
  asUint,
  bytes,
  CborValue,
  isBytes,
  isText,
  isUint,
  text,
  uint,
} from './cbor';
import { Envelope } from './envelope';
import * as op from './op';
import * as blobs from './blobs';

// ============================================================ HELLO (§8.1)

export function helloRequest(
  l1Version: number,
  walletPubkey: Uint8Array, // 32-byte X25519
  supportedOps: number[],
): Envelope {
  if (walletPubkey.length !== 32) throw new Error('wallet_pubkey must be 32 bytes');
  return Envelope.of(
    op.HELLO,
    uint(l1Version),
    bytes(walletPubkey),
    array(...supportedOps.map(o => uint(o))),
  );
}

export interface HelloResponse {
  l1Version: number;
  gatewayPubkey: Uint8Array; // 32-byte X25519
  mintUrls: string[];
  nut19CachedPaths: string[];
  maxMsg: bigint;
  serverTime: bigint;
}

export function parseHelloResponse(env: Envelope): HelloResponse {
  requireOp(env, op.responseOf(op.HELLO));
  const f = env.fields;
  return {
    l1Version: Number(asUint(f[0]!, 'l1_version')),
    gatewayPubkey: asBytes(f[1]!, 'gateway_pubkey'),
    mintUrls: textList(f[2]!, 'mint_urls'),
    nut19CachedPaths: textList(f[3]!, 'nut19_cached_paths'),
    maxMsg: asUint(f[4]!, 'max_msg'),
    serverTime: asUint(f[5]!, 'server_time'),
  };
}

// ============================================================ KEYSETS (§8.10)

export function keysetsRequest(mintHandle: number): Envelope {
  return Envelope.of(op.KEYSETS, uint(mintHandle));
}

export interface KeysetEntry {
  shortId: Uint8Array; // 8 bytes
  unit: string;
  active: boolean;
  inputFeePpk: bigint;
  finalExpiry: bigint;
}

export function parseKeysetsResponse(env: Envelope): KeysetEntry[] {
  requireOp(env, op.responseOf(op.KEYSETS));
  return asArray(env.fields[0]!, 'keysets').map(row => {
    const e = asArray(row, 'keyset entry');
    return {
      shortId: asBytes(e[0]!, 's_id'),
      unit: unitFromCbor(e[1]!),
      active: asUint(e[2]!, 'active') !== 0n,
      inputFeePpk: asUint(e[3]!, 'input_fee_ppk'),
      finalExpiry: asUint(e[4]!, 'final_expiry'),
    };
  });
}

// ============================================================ KEYS (§8.11)

export function keysRequest(
  keysetHandle: number,
  fromExp: number | null,
  toExp: number | null,
): Envelope {
  return Envelope.of(
    op.KEYS,
    uint(keysetHandle),
    fromExp === null ? { t: 'null' } : uint(fromExp),
    toExp === null ? { t: 'null' } : uint(toExp),
  );
}

export interface KeysResponse {
  shortId: Uint8Array;
  keys: Uint8Array[]; // index i = pubkey for amount 2^i (ranged!)
}

export function parseKeysResponse(env: Envelope): KeysResponse {
  requireOp(env, op.responseOf(op.KEYS));
  const f = env.fields;
  // KEYS responses may be ranged: n×33, not fixed 64 (gateway packs directly).
  return { shortId: asBytes(f[0]!, 's_id'), keys: blobs.unpackSignatures(asBytes(f[1]!, 'keys_blob')) };
}

// ============================================================ CHECKSTATE (§8.9)

export const STATE_UNSPENT = 0;
export const STATE_PENDING = 1;
export const STATE_SPENT = 2;

export function checkstateRequest(ys: Uint8Array[]): Envelope {
  return Envelope.of(op.CHECKSTATE, bytes(blobs.packSignatures(ys)));
}

export function parseCheckstateResponse(env: Envelope, proofCount: number): number[] {
  requireOp(env, op.responseOf(op.CHECKSTATE));
  return unpackStates(asBytes(env.fields[0]!, 'packed_states'), proofCount);
}

/** 2 bits per proof, LSB-first (§8.9). */
export function packStates(states: number[]): Uint8Array {
  const out = new Uint8Array(Math.ceil((states.length * 2) / 8));
  states.forEach((s, i) => {
    if (s < 0 || s > 2) throw new Error(`state out of range: ${s}`);
    const idx = (i * 2) >> 3;
    out[idx] = (out[idx] ?? 0) | (s << ((i * 2) & 7));
  });
  return out;
}

export function unpackStates(packed: Uint8Array, count: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < count; i++) {
    out.push((packed[(i * 2) >> 3]! >> ((i * 2) & 7)) & 0x3);
  }
  return out;
}

// ============================================================ SWAP (§8.13)

export function swapRequest(
  keysetHandle: number,
  inputs: blobs.Proof[],
  outputs: blobs.Output[],
): Envelope {
  return Envelope.of(
    op.SWAP,
    uint(keysetHandle),
    bytes(blobs.packProofs(inputs)),
    bytes(blobs.packOutputs(outputs)),
  );
}

export function parseSwapResponse(env: Envelope): Uint8Array[] {
  requireOp(env, op.responseOf(op.SWAP));
  return blobs.unpackSignatures(asBytes(env.fields[0]!, 'signatures_blob'));
}

// ============================================================ RESTORE (§8.8)

export function restoreRequest(
  keysetHandle: number,
  counterStart: bigint | number,
  count: number,
  outputs: blobs.Output[],
): Envelope {
  return Envelope.of(
    op.RESTORE,
    uint(keysetHandle),
    uint(counterStart),
    uint(count),
    bytes(blobs.packOutputs(outputs)),
  );
}

export interface RestoreResponse {
  hitBitmap: Uint8Array; // bit i set → counter (start+i) signed
  signatures: Uint8Array[]; // for set bits only, ascending
}

export function parseRestoreResponse(env: Envelope): RestoreResponse {
  requireOp(env, op.responseOf(op.RESTORE));
  const f = env.fields;
  return {
    hitBitmap: asBytes(f[0]!, 'hit_bitmap'),
    signatures: blobs.unpackSignatures(asBytes(f[1]!, 'signatures_blob')),
  };
}

// ============================================================ ERROR (§8.14, §10)

export interface ErrorResponse {
  code: bigint;
  detailKind: number; // 0 none, 1 text, 2 bytes
  detailText: string | null;
  detailBytes: Uint8Array | null;
}

export function error(code: bigint | number): Envelope {
  return Envelope.of(op.ERROR, uint(code), uint(0));
}

export function parseError(env: Envelope): ErrorResponse {
  requireOp(env, op.ERROR);
  const f = env.fields;
  const code = asUint(f[0]!, 'code');
  const kind = Number(asUint(f[1]!, 'detail_kind'));
  let detailText: string | null = null;
  let detailBytes: Uint8Array | null = null;
  if (kind === 1 && f.length > 2 && isText(f[2]!)) detailText = f[2].v;
  else if (kind === 2 && f.length > 2 && isBytes(f[2]!)) detailBytes = f[2].v;
  return { code, detailKind: kind, detailText, detailBytes };
}

// ============================================================ helpers

function requireOp(env: Envelope, expected: number): void {
  if (env.opcode !== expected) {
    throw new Error(`expected opcode 0x${expected.toString(16)}, got 0x${env.opcode.toString(16)}`);
  }
}

function textList(v: CborValue, what: string): string[] {
  return asArray(v, what).map(item => {
    if (!isText(item)) throw new Error(`expected text in ${what}`);
    return item.v;
  });
}

/** §6.6: well-known units as uint, custom units as text. */
function unitFromCbor(v: CborValue): string {
  if (isUint(v)) {
    return ['sat', 'msat', 'usd', 'eur', 'btc'][Number(v.v)] ?? `unit:${v.v}`;
  }
  if (isText(v)) return v.v;
  throw new Error('unit must be uint or text');
}
